### ISSUE TITLE
Package Manager and assembly definitions

### ISSUE THREAD

**Issue by metamorphling** [Demigiant/dotween#251]
Are there any plans for moving onto unity's new package management system and usage of assembly definitions? Nothing urgent here, just an enhancement idea.
Tried to move "Demigiant" into "Packages", added packages.json and registered it with manifest - so far works well. Utility Panel is dead though, probably it wants dotween folder to be at hardcoded place?
I'm sure you are really busy, but just for the record dropping a few links on what I am talking about

https://docs.unity3d.com/Packages/com.unity.package-manager-ui@1.7/manual/index.html
https://docs.unity3d.com/Manual/ScriptCompilationAssemblyDefinitionFiles.html

Besides utility panel, as to what I had to do to make assembly definitions work with DOTween:
1) Create core asmdef file in "Modules" folder
Seeing how project went modular way, there is a potential to breakdown every module into it's own folder and give asmdef file for every folder. That way switching modules will be as easy as having a core asmdef which adds/removes references to module assembly definitions.
2) Put scripts from "DOTweenPro" into "DOTweenPro/Scripts" and create pro asmdef file in Scripts
For some reason if I created asmdef in DOTweenPro as is some strange errors were popping up, got the feeling asmdef was fighting dll.
3) Reference both core and pro asmdefs from source code

**Comment by Demigiant** [comment]
Ahoy!

Thanks a lot for the investigation on the matter: I will keep these info as a precious reference :) I'm waiting for Unity 2018.3 which will finally make the package manager available to everyone, and then I'll implement it for sure.

Cheers,
Daniele

**Comment by metamorphling** [comment]
> It is highly recommended that you use assembly definition files for all the scripts in the Project, or not at all. Otherwise, the scripts that are not using assembly definition files always recompile every time an assembly definition file recompiles. This reduces the benefit of using assembly definition files.

Important point, to support both methods it would be cool to have a checkbox in settings menu. Like, "Use Assembly Definitions". Since .asmdef files are plain text files, they can be generated and deleted with ease,

**Comment by laicasaane** [comment]
I'm maintaining a multiple projects setup that has a copy of DOTween in each sub project just because of the utility panel problem. So I hope this feature would be done the sooner the better. No pushing from me, though, just hope so.

**Comment by Demigiant** [comment]
I plan to finally experiment with asmdef today, but not on Package Manager to be honest :B It still seems very alpha so I'm verging on the side of "let's postpone that more".

@laicasaane can you explain me more? I'm not sure I understood everything of the setup and of the problem there :B

**Comment by laicasaane** [comment]
@Demigiant My project has been divided into 3 sub projects inside it, 1st is the core, 2nd is for the players, 3rd is for the content makers. All of them use DOTween and other libraries. Others can be made into local packages just fine, but DOTween couldn't. Actually I had reported that issue here https://github.com/Demigiant/dotween/issues/238

I know there wouldn't be any issue in runtime, and this only affects the Unity editor. But I really don't like errors and exceptions appear even in the editor, so I've to put a copy of DOTween in each sub project to avoid this exception.

**Comment by metamorphling** [comment]
I too had some problems trying to isolate dotween into a package for multiple projects environment. Don't have a playground at hand, but if I remember correctly the cause of the problem was something like hardcoded paths for some elements of the plugin. As in, trying to access files(settings or editor images?) from the "Assets/smthsmth" while real path for package is under "Packages/PackageName/Assets/smthsmth". Setting paths to resolve for Assets and Packages respectively, depending on if dotween integrated or plugged with project, should solve the issue. Another possibility of what might go wrong is Resources.Load api, since if there are a few packages/projects with same files it does not differentiate of what to use and things may go wrong.

I feel like this is actually a common problem for plugin makers this days. Unity is halfway onto breaking engine parts into packages, while some users are jumping ship and divide environment too. At this point I can only advice this approach to those who do not use plugins, have no plans to add them midway(which is hard to anticipate) or use them selectively - only tested and confirmed stuff. Meanwhile, let's hope DOTween will be able to support new approach sooner than later ;)

**Comment by Demigiant** [comment]
Ah! Thank you both for the info. I was going to post a new version with auto-asmdef creation which I just finished, but after reading what you wrote I'm going to try and at least implement a way to find DOTweenSettings in any folder, so even if I don't do Package Manager myself it should work for you. Back in a short while, I hope :P

**Comment by Demigiant** [comment]
Ok I created an external DOTween package and tried until now to make it work, but I finally had to ragequit :B I managed to get all references correctly, but the problem is that I also need to modify files via the editor, and it's a complete mess. I experimented but the lack of documentation is too great, so I'll postpone this for the day Unity decides to make Package Manager a little more open, sigh :|

In the meantime, [here's v1.2.225](http://dotween.demigiant.com/tmp/DOTween_1_2_225.zip), where I added a button in DOTween Utility Panel to create the asmdef files. It creates one for the modules and one for DOTweenPro loose scripts (if present). I managed to create DOTweenPro's directly in the DOTweenPro folder by manually adding references to the asmdef json when it's created.

**Comment by LittleBigMonkey** [comment]
So any news on this ?

**Comment by cg-adam** [comment]
Any updates on this? Would be great to have this as a package.

**Comment by Demigiant** [comment]
Ahoy! The documentation remains too intricate (for the things I need to do with DOTween), so until Unity officially allows the implementation of external packages with at least some API I'm postponing this :|

**Comment by LittleBigMonkey** [comment]
Hi ! I'm using Git packages for most plugin I use in every project. It's sooo useful and easier to maintain. It's not too hard to use, you should look at it to make your plugin compatible. It would be awesome to not import it in every single project...

**Comment by Razenpok** [comment]
@mob-sakai made a couple of packages that are available via git
https://github.com/mob-sakai/SoftMaskForUGUI
https://github.com/mob-sakai/ParticleEffectForUGUI
https://github.com/mob-sakai/UpmGitExtension

**Comment by Demigiant** [comment]
Mhmm ok you're convincing me to give this another try :P But I already know this will require quite a lot of time (I'll have to change the editor initialization logic of DOTween and other systems). I'm in crunch right now so I'll see if I can start working on this by the middle of next week.

**Comment by jasursadikov** [comment]
@Demigiant any news on this?

**Comment by hoseongcheon** [comment]
@Demigiant any news on this?

**Comment by Demigiant** [comment]
I come back to it now and then, but always encounter new problems ([see last one here](https://github.com/Demigiant/dotween/issues/344#issuecomment-553309397)). I definitely intend to work more on it but too much of DOTween is mutable and built with the options you choose via DOTween Utility Panel's modules, so it doesn't seem a good fit for the Package Manager. The alternative would be to make it less mutable like it was before, but I think that is a much greater advantage than PM.

Sooo, in short, for now there's no date for when this might happen, and it might not happen :|

**Comment by LittleBigMonkey** [comment]
I don't know why someone would desactivate basic unity modules. The only one I don't care, is 2D Toolkit.

But anyway, instead of having one package that is mutable, why don't you try to have multiple package and include them in the package manager when needed. You could have something like :
- DoTween.Core
- DoTween.Modules.UI
- …

That's how unity manage their own modules in the packages manager. And we can simply activate/desactivate a module by adding/removing a line in the manifest. Moreover, you can specify which revision you want if you need to support multiple unity version.

**Comment by Demigiant** [comment]
The separation into basic Unity modules was a feature that required a lot of work, which I added after some users asked it :B It's rather important because if, for example, you have no Physics in your game and no script referring it then Unity doesn't include the Physics system in builds, so it helps making builds slimmer by removing stuff you're not using.

The multiple packages idea sounds neat, I'm going to try that out :)

**Comment by laicasaane** [comment]
I've deployed a customized repo so I can use DOTween with Unity package management system. And it's working almost fine so far. The only thing broken is DOTween Utility Panel. The panel won't show, and the DOTweenSettings.asset is read-only. At the moment, this is my strategy to solve that problem: edit the settings within an empty project, then copy that settings asset file over the main project.

![image](https://user-images.githubusercontent.com/1594982/79744986-88cd1800-8331-11ea-877c-6f7d07afe79f.png)

Really hope to use DOTween via UPM officially one day.

**Comment by Demigiant** [comment]
Sadly it's even more complicated than that. The Modules system changes actual DOTween files, so even if that would work it would overwrite all Modules-based DOTween scripts for all projects using the package, which is not nice :| I hope one day to give this more time (because it's gonna be a long and major refactoring) but I'm prioritizing other things for now (like a DOTweenTimeline which I've been working on for the last months :P)

**Comment by laicasaane** [comment]
Beside that feature you are working on, could you prioritize making the settings accessible no matter where it is? IMO this is the only problem urgently needs to be fixed. The modules system could wait.

**Comment by Demigiant** [comment]
I never really thought of implementing it with a "disabled Modules" option. I will think about it but it is kind of sad.

**Comment by laicasaane** [comment]
For the time being, I think you should let all modules turned on by default, and let people know that if we use DOTween as a package we won't be able to turn off those modules, and that should be the limitation at this point. I think since all modules are on by default, only in some rare cases people will run into any problem and have to import DOTween directly into their project.

This behaviour is more relax than the current one: the settings that is not editable directly in the project is really annoying.

**Comment by bdovaz** [comment]
@Demigiant If you implement UPM you could resolve this better.

Check UniTask repo on how implement UPM support:

https://github.com/Cysharp/UniTask#upm-package

Then you could separate in more packages.

com.demigiant.dotween
com.demigiant.dotween.ugui
com.demigiant.dotween.physics
...

And we would use the packages we want.

Regarding DOTween *.asmdef support I have a problem and is that if I enabled it it would generate a different *.meta with a different GUID. I have my packages depending on DOTween by GUID so in different projects your *.asmdef GUID is different and it brokes our references.

**Comment by mschweitzer-sd** [comment]
> I don't know why someone would desactivate basic unity modules. The only one I don't care, is 2D Toolkit.
> 
> But anyway, instead of having one package that is mutable, why don't you try to have multiple package and include them in the package manager when needed. You could have something like :
> 
> * DoTween.Core
> * DoTween.Modules.UI
> * …
> 
> That's how unity manage their own modules in the packages manager. And we can simply activate/desactivate a module by adding/removing a line in the manifest. Moreover, you can specify which revision you want if you need to support multiple unity version.

> @Demigiant If you implement UPM you could resolve this better.
> 
> Check UniTask repo on how implement UPM support:
> 
> https://github.com/Cysharp/UniTask#upm-package
> 
> Then you could separate in more packages.
> 
> com.demigiant.dotween
> com.demigiant.dotween.ugui
> com.demigiant.dotween.physics
> ...
> 
> And we would use the packages we want.
> 
> Regarding DOTween *.asmdef support I have a problem and is that if I enabled it it would generate a different *.meta with a different GUID. I have my packages depending on DOTween by GUID so in different projects your *.asmdef GUID is different and it brokes our references.

+1 to both of these suggestions. TBH it feels like users asked for the DOTween modules approach because it was basically a workaround for the lack of a proper package management solution. Now that upm exists, I think the DOTween modules implementation is no longer necessary and should be replaced with separate packages that depend on a core package. Users can then decide via the package manager which modules they want.

It seems like the main blocker is the fact that specific ScriptableObjects are assumed to be in specific locations in a project. I think this is generally a risky approach since it makes lots of assumptions about project layout. A suggestion I have is to invert the dependencies - either pass in references to the ScriptableObjects when you need them, save them off as properties so they can be set independently, or provide callbacks to let the user supply the objects. As far as the utility panel, maybe you just don't show anything until the user drags in or provides a location for the Settings object.

I know that this would be a major refactoring, but it's worth it. upm is here to stay and all of Unity's engine components are being converted to use it. I think the comments blaming Unity for lacking documentation are unfair honestly. The documentation is quite good IMO and lots of plugins are starting to refactor to support upm (NaughtyAttributes, Deform, etc).

**Comment by pshtif** [comment]
We are just in the middle of packaging all our plugins and modules through UPM as well and actually the only 3rd party asset that can't be used through UPM seems to be DOTween. At this moment it can get so critical we will be probably looking to implement our own tweening library rather than using DOTween the old fashioned way, which in a big company was clunky to begin with. So we are welcoming Unity's modular iniciative and UPM and hopefully DOTween can join the bandwagon.

**Comment by leon-arndt** [comment]
Huge DOTween fan, would love to use it as a package as well. Any update on this? Unity move towards packages seems to be clear.

**Comment by leandro-manifesto** [comment]
Seems like someone hijacked (maybe not intetionally) the DOTween package on OpenUPM: https://openupm.com/packages/com.demigiant.dotween/

**Comment by laicasaane** [comment]
@leandro-manifesto 
The problem is there is no way we can edit DOTweenSettings even if there is a copy in Resources folder

![image](https://user-images.githubusercontent.com/1594982/89856428-75d1a280-dbc3-11ea-8450-392d847a7344.png)

**Comment by Demigiant** [comment]
I don't know what the hijack means :B
I've continued to make tests and discuss the issue with other devs (since by now I gave up hoping that UPM will get better and more dev-friendly) and I think I have a way (inspired from your suggestions of separating modules in packages), but it's a really complicated way that will take a while to be implemented and I'm giving priorities to other DOTween stuff first (like the Timeline, if you heard about it). I really have no timeframe for this. After I release the Timeline I will probably move to this though.

**Comment by edwonedwon** [comment]
Is this the latest news on Unity Package Manager support? I would love this!

**Comment by JimmyCushnie** [comment]
I too would really love this :)

**Comment by Zanleo** [comment]
Hi, I ran into a big need in Package for DoTween. Since a large number of internal developments are tied to DoTween and are delivered via PackageManager, but DoTween does not work from the package (

**Comment by edwonedwon** [comment]
I need this so much as well! I have several of my own packages that have dependencies on DOTween

**Comment by jansensevr** [comment]
We really need DoTween in UPM. We are currently working on dividing our project into smaller bits so the production team can more easily stay up to date with internal editor tools etc and ofc some of the editor code is dependent on DoTween. Since DoTween doesn't work when we put it in a package we cannot have any dependencies on it in our packages :/

Tomorrow I will try another solution as I believe we can manually wrap DoTween into an assembly definition and then reference that by name in the package. We will still need to manualy import DoTween into each project, but hey! it's still better than not having it at all. At least this modified version of DoTween can be wrapped into a UnityPackage and since we would reference the assemblydef by name from our packages, as far as I can tell, a simple unitypackage import should do the trick. Hope that works :)

**Comment by sicklydove** [comment]
Another +1 for this here. _Big_ fan of DoTween, and have been using it for some functionality bundled up into a Package for a client. Unfortunately, I've had to remove it and reimplement my own basic Tweening with Unity's APIs, as this package cannot require any extra install steps. Would be really great to be able to use this.

**Comment by rhys-vdw** [comment]
> I think I have a way (inspired from your suggestions of separating modules in packages)

I just wanted to add my thoughts to this thread (as I have [here](https://github.com/Demigiant/dotween/issues/395#issuecomment-823023631)). You don't need to use multiple packages, you can have all the modules in the _same_ package, but have each in its own `asmdef` which can be selectively imported into your project (and therefore included in your builds). I believe [UniTask](https://github.com/Cysharp/UniTask) is probably the best reference for package architecture, it includes extensions that do not even _compile_ unless its dependent packages are found (very cool, but not necessary for DOTween's packages). For example it includes a package of DOTween extensions that only compile if your project includes the aforementioned openUPM DOTween package.

**Comment by mschweitzertest** [comment]
Agree with @rhys-vdw, you can have a single upm package and split it into multiple asmdefs using the same strategy that was used to split the old DOTween modules. The upside is less packages to manage, but the downside is that it makes the Unity project itself bigger since the package will still download the source even if it's unused.

I think at this point whatever gets upm support faster is the right way to go.

**Comment by mikerochip** [comment]
Agree with @rhys-vdw, you can have a single upm package and split it into multiple asmdefs using the same strategy that was used to split the old DOTween modules. The upside is less packages to manage, but the downside is that it makes the Unity project itself bigger since the package will still download the source even if it's unused.

I think at this point whatever gets upm support faster is the right way to go.

**Comment by rhys-vdw** [comment]
The larger download is not a big deal, it's just a few text files and they don't even go in your project's version control. There are really no downsides and it means no more need for this custom module system and  setup screen.

**Comment by emathis11** [comment]
Why not just use Version Defines (in the asmdef configuration)?

For example if the Unity built-in package "Physics2D" is installed (com.unity.modules.physics2d) it will define a compilation symbol such as "USE_PHYSICS_2D" that you can then use to wrap code that uses Physics2D, so that it is ignored if the user has disabled the Physics2D package.

The downside is that the user has to disable the whole Physics2D module in Unity instead of just the DoTween module. But this seems like the right approach with UPM.

**Comment by fwalker007** [comment]
Is there a solution for this anywhere? We really need this as well.

**Comment by rhys-vdw** [comment]
@fwalker007 what's the problem? I'm using the package on OpenUPM and it works okay.

**Comment by bdovaz** [comment]
@rhys-vdw in my case I can't change DOTweenSettings.asset via editor window. It freezes loading the data.

**Comment by rhys-vdw** [comment]
@bdovaz Yeah, I have the same thing but there are no negative consequences.

**Comment by bdovaz** [comment]
@rhys-vdw yes, we can still change this via global static settings.

**Comment by ExodusOTH** [comment]
Another +1 :) Whenever you get around to working on it, it'd be a big help to me too, I've grown to really like DOTween but I'm having to start shave away any dependencies not using the package system for my own personal toolkit

**Comment by theReallyRealDusty** [comment]
Any news on this? Moved to asmdefs and had a bunch of these problems. 
`Assets\Plugins\Demigiant\DOTween\Modules\DOTweenModulePhysics.cs(188,65): error CS0411: The type arguments for method 'DOTween.To<T1, T2, TPlugOptions>(ABSTweenPlugin<T1, T2, TPlugOptions>, DOGetter<T1>, DOSetter<T1>, T2, float)' cannot be inferred from the usage. Try specifying the type arguments explicitly.`
Tried putting the dotween dlls in the assembly references by ticking the "Override references" in the assembly definition settings and it does not seem to do anything. This approach works fine with Odin which also has precompiled code.
Quite a shame as I had to remove all references to DOTween for now. Luckily it was only 2 places since I didn't really get to the animate and make everything very pretty part.

**Comment by rhys-vdw** [comment]
@lorinatzberger The package on OpenUPM is massively out of date. I set up DOTween this week using the downloaded version. The main reason I had for using the package is to enable to UniTask extensions (which are disabled by default unless you have the package manager version included). My solution to this was copying the UniTask DOTween extension out into the main package.

@Demigiant would you be open to me opening a PR to support UPM? I have done this for [ink-unity-integration](https://github.com/inkle/ink-unity-integration) and for my own package [unity-destringer](https://github.com/rhys-vdw/unity-destringer). I can't guarantee to deliver it immediately, but I am confident that I'd do a good job of it. Maybe in the next few weeks.

One potential blocker is that @yoshida190 has taken your namespace/package ID `com.demigiant.dotween`, despite not owning demigiant.com. Hoping they will respond here as they have issues disabled on [their fork](https://github.com/yoshida190/dotween).

**Comment by theReallyRealDusty** [comment]
@rhys-vdw Even if it would be available on OpenUPM, I don't think that'll work with the pro version since then, there would be no reason to actually buy it. I'm talking about the latest version available in the asset store through "My assets" in the regular package manager.

**Comment by Razenpok** [comment]
@rhys-vdw 

Regarding the taken package ID - I'm pretty sure you can just ask @favoyang (the creator of OpenUPM) to transfer the ownership if needed.

**Comment by favoyang** [comment]
> One potential blocker is that @yoshida190 has taken your namespace/package ID `com.demigiant.dotween`, despite not owning demigiant.com. Hoping they will respond here as they have issues disabled on [their fork](https://github.com/yoshida190/dotween).


> Regarding the taken package ID - I'm pretty sure you can just ask @favoyang (the creator of OpenUPM) to transfer the ownership if needed.

I have received a request from @yoshida190 to retire his fork from OpenUPM, as long as anyone like to provide a replacement, either the official one or a maintainable fork. Please [create an issue](https://github.com/openupm/openupm/issues/new/choose) on OpenUPM to propose your alternatives.

**Comment by rhys-vdw** [comment]
> @rhys-vdw Even if it would be available on OpenUPM, I don't think that'll work with the pro version since then, there would be no reason to actually buy it. I'm talking about the latest version available in the asset store through "My assets" in the regular package manager. 

I don't know anything about the pro version. I assume it's not open for pull requests. But generally speaking there's no reason why the exact file structure of a well structured unity package can't be directly dropped into a unity project, so whatever changes are made can be applied to both.

Fwiw I am using asmdefs with latest version on GitHub and it's working perfectly.

**Comment by rhys-vdw** [comment]
> I have received a request from @yoshida190 to retire his fork from OpenUPM, as long as anyone like to provide a replacement, either the official one or a maintainable fork. 

Thanks @favoyang.

I think maintaining a fork would not be ideal here as ultimately DOTween itself should be updated to support modern development practises.

Also I suspect it will be hard to automate the updating process, and setting up CI is a bit more than I'm signing up for here.

I'll wait for @Demigiant to consider my proposal.

**Comment by theReallyRealDusty** [comment]
So, any news on this? I understand that it may be a more complicated situation but it's been 2 whole years. As it is, I can't even run the setup since it throws errors right after importing.

**Comment by theReallyRealDusty** [comment]
Yeah, giving up on making this run again.
I tried deleting anything about the pro version to simplify things then making 4 asmdefs for DemiEditor, DemiLib, DoTweenEditor(including both DoTweenEditor and DoTweenUpgradeManager) and DoTween. I set the correct editor settings and used Override References to link the asmdefs to their respective dlls.

All I get is the console spamming things constantly:
`TypeLoadException: Could not load type of field 'DG.DOTweenEditor.UI.DOTweenUtilityWindowModules:_src' (14) due to: Could not resolve type with token 01000012 (from typeref, class/assembly DG.Tweening.Core.DOTweenSettings, DOTween, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null) assembly:DOTween, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null type:DG.Tweening.Core.DOTweenSettings member:(null) signature:<none>
System.RuntimeType.GetMethodsByName (System.String name, System.Reflection.BindingFlags bindingAttr, System.Boolean ignoreCase, System.RuntimeType reflectedType) (at <695d1cc93cca45069c528c15c9fdd749>:0)
System.RuntimeType.GetMethodCandidates (System.String name, System.Reflection.BindingFlags bindingAttr, System.Reflection.CallingConventions callConv, System.Type[] types, System.Boolean allowPrefixLookup) (at <695d1cc93cca45069c528c15c9fdd749>:0)
System.RuntimeType.GetMethodImpl (System.String name, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Reflection.CallingConventions callConv, System.Type[] types, System.Reflection.ParameterModifier[] modifiers) (at <695d1cc93cca45069c528c15c9fdd749>:0)
System.Type.GetMethod (System.String name, System.Reflection.BindingFlags bindingAttr) (at <695d1cc93cca45069c528c15c9fdd749>:0)
DG.DOTweenUpgradeManager.Autorun.ApplyModulesAndASMDEFSettings () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTweenUpgradeManager/Autorun.cs:44)
DG.DOTweenUpgradeManager.Autorun.OnUpdate () (at D:/DG/_Develop/__UNITY3_CLASSES/_Holoville/__DOTween/_DOTween.Assembly/DOTweenUpgradeManager/Autorun.cs:30)
UnityEditor.EditorApplication.Internal_CallUpdateFunctions () (at <01e901910c5242498b7e69f6d2d8d500>:0)
`

I also did somehow managed to get the setup window to show at some point but clicking on the setup button just threw an error.

It's not even about the fact that it should necessarily work out of the box but a guide would be really nice.

**Comment by krisrok** [comment]
@Demigiant daniele, help us out here :)

**Comment by Maestroxr** [comment]
*crickets*

**Comment by rhys-vdw** [comment]
Still willing to do this one if approved.

Alternatively I could look into making a fork that uses github actions to pull in upstream changes? I'd need a bit of time to research but it could be the best stopgap while we wait for official support.

**Comment by Demigiant** [comment]
Hey!
Sorry for the radio silence here but every time I see a UPM notification my mind explodes, then I make plans to try doing it again, then either try and fail or cancel the idea; and lately I let these messages go over my head (so much that I missed the fact that the com.demigiant.dotween namespace had been taken, which is not great).
I tried to implement it many times in the past, and in various ways, but always had show-stopping problems: both because of the modules system that needs to re-write files and because anyway that would never work with the Pro. I humbly ask you to please continue using the UPM-AssetStore side of it, or if you really need to include stuff for clients/projects you can simply include DOTween's files in the project (that is perfectly allowed, as long as it's not the Pro)? I know it's nicer to have it working directly from UPM, but it's not that bad to have to go through the Asset Store, or am I missing something obvious? Honest question, agh 👀💦

**Comment by rhys-vdw** [comment]
Thank you for replying @Demigiant!

_sorry for long message but I want to make my case_

I really love this library and it bugs me that it's not using up-to-date practises.

> because of the modules system that needs to re-write files

Actually there is support for this in UPM. (I explained above.) Basically Unity will set compile flags based on presence of other dependencies, so this whole custom module system could be removed and replaced with the native support. Your module system, while clever, is made redundant by this Unity feature. (actually this is one of the things I personally would most like to remove)

> anyway that would never work with the Pro

Actually it's completely fine. Customers can put your paid package right into the `Packages/` folder in the project root. Unity will treat it as local UPM package. I've seen other paid assets that are distributed this way.

(I believe you can also just drop a valid UPM package into `Assets/` and it will work too, but it's probably fine to drop support for pre-UPM versions of Unity at this stage.) 

> I missed the fact that the com.demigiant.dotween namespace had been taken, which is not great

See [above](https://github.com/Demigiant/dotween/issues/251#issuecomment-899499535) where OpenUPM maintainer confirms that you can take back the namespace if an official release is made.

> I know it's nicer to have it working directly from UPM, but it's not that bad to have to go through the Asset Store, or am I missing something obvious?

From my perspective there are a bunch of things that could be nicer:

1. Upgrading to latest release (or maintaining a fork) is complicated because you can't just change the reference in package manager.
1. It's hard/impossible to work out which version of DOTween you have installed in a project. (Whereas the UPM dependency would just tell you in the `packages.json`.)
1. Currently it's not possible to distribute optional extensions for DOTween via UPM. UniTask has a suite of DOTween extensions. It uses the Unity dependency system to enable its extensions based on the presence of the `com.demigiant.dotween` package. These extensions _cannot be used_ without relying on the old, bad, broken fork of DOTween, or manually copying the UniTask extensions into your assets.
1. Distributing straight from git/OpenUPM makes source easily accessible.
1. Upgrading/downgrading a UPM package does not have the problem of old renamed or deleted files remaining, it always does a full immutable copy of the version you want.
1. As a consumer it's preferable for dependencies to not be included under version control. This prevents accidental modification of dependencies and repo bloat.
1. The current implementation for some reason deletes any asmdef that is added to the plugins folder, which means that you cannot use your modules _at all_ if you're using asmdefs, as there's no way to reference them from your assembly! (this is a separate issue but damn it's annoying)

> I tried to implement it many times in the past, and in various ways

As I mentioned above I've done a few of these now and have researched how to make the DOTween module system work. I'm confident that I can do it. Obviously you'd need to merge the changes to the pro version as I have no repo access.

**Comment by RunninglVlan** [comment]
I like UPM packages too, but what if the package user wants to modify it, e.g., to make a fix or extend its functionality? Using Asset Store packages we could just make a modification, but how to do it using UPM?
Do we need to copy it first to /Packages? Or make a fork and modify it instead?

**Comment by rhys-vdw** [comment]
> Do we need to copy it first to /Packages? Or make a fork and modify it instead?

These are the two options, yes.

Or if you don't share my concerns you can just download the asset store version. This would be for people who want the UPM version. The manifest is just ignored if you grab the package from the store (or the dotween website which is where I get it from)

Just to be completely clear, this would _add_ the ability to distribute via UPM, not make DOTween incompatible with store or private distribution (for the pro version). That would be unacceptable as I'm sure the majority of users (esp. beginners) will still be getting it from the asset store.

**Comment by mikerochip** [comment]
@Demigiant If you're finding yourself trying to do the work and giving up, then just let @rhys-vdw do it. That's the beauty of being open source on git and hosting on GitHub. The problem is that nobody besides yourself has repo access. Here's how you can do it:

1. Create a branch called ```upm``` and push it
2. Let @rhys-vdw make the code UPM compatible, then he can open a Pull Request against the ```upm``` branch
3. You accept and merge the Pull Request

That's it. No heavy lifting for you and everyone that wants UPM support benefits.

With this method, you can leave all your asset store stuff as it is. This GitHub repo would just provide additional upm support via the branch.

**Comment by Demigiant** [comment]
Since it's the holidays, let's keep this as a conversation for now, so I can understand better and make up my mind :)

@rhys-vdw  I understand your point of view but I have a few questions/doubts:

> Actually there is support for this in UPM. (I explained above.) Basically Unity will set compile flags based on presence of other dependencies, so this whole custom module system could be removed and replaced with the native support. Your module system, while clever, is made redundant by this Unity feature. (actually this is one of the things I personally would most like to remove)

The problem is that, yes, I know that, but that would work only in recent versions of Unity, while I always strived to keep DOTween as retro-compatible as possible, because there's lots of teams still working on astoundingly old versions of Unity (just a couple weeks ago I had an email from a team which had problems updating because the recent versions of DOTween require Unity 5.6 while they were still on 5.4). So the self-rewriting modules are imo the best way (other than having DOTween add custom defines, like it did before, but which is a practice I really didn't like because it was too invasive). This is also why I see the UPM approach as limited: I don't think Unity gave it the love it deserved, and it's only for recent versions.

> It's hard/impossible to work out which version of DOTween you have installed in a project. (Whereas the UPM dependency would just tell you in the packages.json.)

Can you explain me better here? Afaik UPM doesn't add custom defines, but maybe I'm missing something. Right now you can determine DOTween's version visually by opening DOTween's Utility Panel, or code-wise by querying DOTween.Version

> Upgrading/downgrading a UPM package does not have the problem of old renamed or deleted files remaining, it always does a full immutable copy of the version you want.

This I love, because the leftover files is one of the main reason I made DLLs instead of loose scripts

> The current implementation for some reason deletes any asmdef that is added to the plugins folder, which means that you cannot use your modules at all if you're using asmdefs, as there's no way to reference them from your assembly! (this is a separate issue but damn it's annoying)

Agh! Can you explain me more? I'm not sure I understand: ASMDEFs are working on my side

@mikerochip 

> If you're finding yourself trying to do the work and giving up, then just let @rhys-vdw do it. That's the beauty of being open source on git and hosting on GitHub. The problem is that nobody besides yourself has repo access.

Your words are wise indeed, but I always preferred to keep direct control of the repo because of many reasons. That said, I will ponder this seriously while we have this discussion.

**Comment by Arcnor** [comment]
>> If you're finding yourself trying to do the work and giving up, then just let @rhys-vdw do it. That's the beauty of being open source on git and hosting on GitHub. The problem is that nobody besides yourself has repo access.
> 
> Your words are wise indeed, but I always preferred to keep direct control of the repo because of many reasons. That said, I will ponder this seriously while we have this discussion.

Just my 2c as an outsider "lurker" interested in this issue for a long time: I don't think the posters above meant you giving or even losing control of the repo, but working on a pull request basis, which still needs your approval/merging before any changes happen, but those can start happening independent from you, which is the good thing, I think :).

**Comment by rhys-vdw** [comment]
_Sorry yet again for another short essay! Don't feel there's an time pressure on responding_

> Since it's the holidays, let's keep this as a conversation for now, so I can understand better and make up my mind :)

Hey @Demigiant. Yeah for sure. I think it's a complicated problem to solve and were you to accept my help I'd need to ask you questions along the way, as well as make a proposal for your approval before I start. My main concern is doing a bunch of work and it not being appropriate. I wouldn't want to waste your or my time, nor put you in a position where you felt obligated to accept work that you do not feel comfortable with.

Also I originally made the offer a while back where I had a chunk of time available, but I will find time to do it still.

> ...I always strived to keep DOTween as retro-compatible as possible...

Definitely a problem to grapple with, I'll keep that in mind. Might need a bit more thinking. Older versions of Unity will simply ignore asmdefs, which should afford some freedom.

If you hadn't mentioned that you objected to adding defines then I might suggest telling consumers to add them manually or create a simple wizard that does it. It is a bit annoying (I've had dependencies that do it). I'd say it's a small price to pay for running a long unsupported version of Unity and still getting updates, but I'm sure there's a way to keep the current module system for `.unitypackage` releases if that is determined to be the best option.

I don't know the details of how you support older Unity versions, or how your build + deploy pipeline works. Probably worth mentioning that I'd need to have a proper read of the codebase before I can lay out what compromises must be made (if any).

> Can you explain me better here? Afaik UPM doesn't add custom defines, but maybe I'm missing something. Right now you can determine DOTween's version visually by opening DOTween's Utility Panel, or code-wise by querying DOTween.Version

Ah cool. I had never considered that. I have just looked in the README in the past to find a version because that's the place I thought to look, but the version does not seem to be in there. So instead I am just vigilant with including the installed version in git so I can search back through my history for it.

I guess just as a comment it's easier to have a standard way of checking for updates (open up package manager and it tells you the version and whether an update is available, click once and you're up to date) rather than looking for each dependency's version and cross-referencing it with the website/asset store, deleting the existing folder, downloading the new package, extracting it etc.

> the leftover files is one of the main reason I made DLLs instead of loose scripts

Yeah, I had a suspicion that that was why. It's an annoying flaw of the old system.

>  Can you explain me more?

Okay, so I just ran into this two weeks ago integrating the package with a new project.

In my case I have `Assets/_Game/Game.Runtime.asmdef` which contains my projects code (I think this is pretty standard). This asmdef can only see code in DLLs or asmdefs that are explicitly referenced by it. There is no asmdef in `Assets/DOTween/Modules`, presumably because it assumes the old non-asmdef system where all code is grouped into the main assembly. This means that my assembly can't use the modules (regardless of module system changes).  So typically the solution is you add an asmdef to the folder so that you can include it in your project.

However when I add in an asmdef it's immediately deleted with this confusing error:

![image](https://user-images.githubusercontent.com/822105/147243278-9fc5af68-fa3a-45bb-a560-c134836413f0.png)

As a result I needed to copy out the modules and put them somewhere else where your library doesn't do this. It's very confusing. I would argue that this kind of automatic self-modification is as invasive as a helper that modifies your build defines, but perhaps I don't have the full picture here.

(In any case the obvious fix here is to add the `asmdef` in the distributed package, I just mention it as an example of non-standard and confusing behaviour)

> I always preferred to keep direct control of the repo

I take this to mean that you mean that you do most of the bug fixes and features yourself? I understand.

I saw the popularity of this issue and thought "hey, it's a big ask of the maintainer to do this work, and I think I could do this". I like the idea of the challenge and contributing to a great library. I understand what it is to have a codebase you're proud of, and won't be offended if you don't take me up on it.

Alternatively I'd be happy to take a look at help formulate a plan for you to execute without me, or honestly to just continue using it in its current form and despite my (admittedly minor) packaging quibbles.

Hope that helps. Happy holidays!

**Comment by rhys-vdw** [comment]
As a followup I have discovered that if I name the asmdef by a different name it doesn't get deleted. I'm guessing you'll have a better understanding of why this happens. Quick search suggests it's coming from [here](https://github.com/Demigiant/dotween/blob/90093a5c775ea22c9b3a2c828557f22a743c2589/_DOTween.Assembly/DOTweenEditor/ASMDEFManager.cs#L172-L198).

**Comment by krisrok** [comment]
> However when I add in an asmdef it's immediately deleted with this confusing error

Have you tried generating asmdefs via the DOTween setup dialog? Maybe there‘s a watcher only allowing asmdefs created that way for some reason?

(Great there‘s some motion in here now, yay!)

**Comment by rhys-vdw** [comment]
> Have you tried generating asmdefs via the DOTween setup dialog? Maybe there‘s a watcher only allowing asmdefs created that way for some reason?

Yeah I'm realising now that that's what's going on. No I had no idea, I just saw the package didn't have an asmdef so I added it using the standard naming convention.

**Comment by bdovaz** [comment]
My advice is that if you want to have backward compatibility with versions that do not support *.asmdef files, those developers should be penalized and should be the ones who have to perform steps manually (being documented) and not those who remain in LTS versions of Unity which is the usual.

It doesn't make sense for a team to have an active development, no matter how complex it is in version 5.x today... For 5.4.x we are talking about the year 2016 and we are almost in 2022... Any active development that is not on a supported LTS version is a mistake.

I would solve it in the following way:

For different modules I would split it into different *.asmdef files. Example: DOTween.Ugui, DOTween.Physics, ...

Each of those *.asmdef will have a "version define":

https://docs.unity3d.com/2020.3/Documentation/Manual/ScriptCompilationAssemblyDefinitionFiles.html#define-symbols

Example:

Resource = com.unity.modules.physics.
Define = DOTWEEN_USE_PHYSICS
Version = 1.0.0

And if I remember correctly the actual module system you have created changed a "#if true / false" when editing the configuration window. It would be to change that "#if true" to "#if DOTWEEN_USE_PHYSICS".

This solution would work fine with Unity 5.x instances because you would simply have to manually create the defines in Player Settings for "DOTWEEN_USE_PHYSICS", etc.

If you need help, I'm willing to create the PR without any problem.

**Comment by gromilQaaaa** [comment]
Hello from 2022. Request is still "Open". Its a shame...

**Comment by Romaleks360** [comment]
still waiting for this! go go demigiant we can do this!

**Comment by rhys-vdw** [comment]
> still waiting for this! go go demigiant we can do this!

I think it's safe to assume, at this stage, that it's not going to happen. I've been toying with the idea of making an MIT licensed alternative, but I'm not sure when, if ever, I'll get a chance.

**Comment by laicasaane** [comment]
> I've been toying with the idea of making an MIT licensed alternative

There are already various MIT licensed tweening libraries, should we consider contribute to one of them instead of starting a new one?

**Comment by Demigiant** [comment]
Sorry if I'm keeping this in such low priority, but all the ideas I or other people had are imo either an architectural step backward or don't fully work, and I have to admit that I don't see keeping it as a normal asset as a big problem. That said, I keep this open because sometimes I get back to it (even if, yes, very low priority at this point: package manager drives me crazy 👀)

**Comment by krisrok** [comment]
> architectural step backward or don't fully work
> package manager drives me crazy 👀)

Maybe at this point you could already start to think about nuget compatibility instead…
Unity teased a bit about it during GDC. So it‘ll take another 4-5 years but then you could be an early adopter ;D

**Comment by bdovaz** [comment]
> and I have to admit that I don't see keeping it as a normal asset as a big problem.

@Demigiant is a big problem when you create a package and you want to have DOTween as a dependency...

Apart from all the advantages of being a package, to mention a few: being able to change version easily just by editing the manifest.json file, having all the assets outside the /Assets folder, the fact that it is a package forces you to use the *.asmdef files with the advantages that entails.

**Comment by krisrok** [comment]
> @Demigiant is a big problem when you create a package and you want to have DOTween as a dependency... 

While all this is true, we have to admit upm is in a shitty state. 
Using dotween as dep in another package does _nothing_ in vanilla upm. You’d have to use openupm-cli or @mob-sakai‘s helper packages (do they even still work in 2021+?). 

So unless Unity fixes upm (will not happen) for many (most?) users it‘d still be manual steps to get the dependencies — whether via upm, git, unitypackage or whatever.
So even though I don‘t like the fact I think @Demigiant still has a point being on the fence about this.

**Comment by bdovaz** [comment]
Of course, being in UPM format by itself is useless if it is not uploaded in an openupm-like npm registry (in which there is already a fork of DOTween in package format for years).

**Comment by rhys-vdw** [comment]
> Of course, being in UPM format by itself is useless if it is not uploaded in an openupm-like npm registry (in which there is already a fork of DOTween in package format for years).
— @bdovaz 

Unfortunately this fork is both broken/incomplete (at my last time of using it) and violates the [license](https://github.com/Demigiant/dotween/blob/develop/LICENSE). I was asked to maintain the fork by @yoshida190 and @favoyang, but declined due to this restriction. Because of this license it's not possible to maintain a community fork. I'm trying to avoid reading the DOTween source so that, if I decide to make my own tweening library, I won't accidentally copy the implementation.

Also specifically you can rely on a UPM formatted package directly from GitHub, OpenUPM is just handy for resolving dependencies (as @krisrok mentions). Having DOTween in a upm package is handy for projects that consume it directly, the abililty to use it as a subdependency is a niche case IMO (although I agree it's a pain point for upm).

**Comment by favoyang** [comment]
Due to the license restriction, the `com.demigiant.dotween` package has been removed from OpenUPM (both from the website and the registry).

To avoid breaking your project, please switch to this git dependency in the packages/manifest.json. The license still allows a fork I think. And using the git dependency is not considered a redistribution.

- git+https://[github.com/yoshida190](https://github.com/yoshida190)/dotween.git#v1.2.320-upm
- git+https://[github.com/yoshida190](https://github.com/yoshida190)/dotween.git#v1.2.420-upm
- git+https://[github.com/yoshida190](https://github.com/yoshida190)/dotween.git#v1.2.632-upm

Learn more about the license restriction:

- https://github.com/yoshida190/dotween/blob/upm/LICENSE

Learn more about how to use manifest.json and the git dependency:

- https://docs.unity3d.com/Manual/upm-manifestPrj.html
- https://docs.unity3d.com/2020.1/Documentation/Manual/upm-git.html

I respect Demigiant's choice for the license. I'm not pushing for a change like MIT.

Further submission using `com.demigiant.dotween`  scope has been [blocked](https://github.com/openupm/openupm/commit/efe54bbd84bdd9b55547ff4d21a3e929994944e1#diff-fcc60177e46f1eba756acf4cb2ede8c8cdbacbeeb26ee398ea5210975004fd49), unless something changes.

----

Update: If using a public fork's git dependency is still considered a redistribution, you can fork the project, turn it into a private fork, then replace the UPM git dependencies with your private fork.

**Comment by bdovaz** [comment]
@Demigiant that is, without an effective replacement of the package in openupm, a package that has been providing a service that has not been officially provided for years has been removed from the repository, leaving thousands of developers without an alternative that does not require making modifications to the project. There are cases in which possibly are third party libraries that depend on that package of yoshida.

Wasn't it better for @Demigiant to give permission somehow to avoid having to remove it?

And finally, are you willing to accept a PR that turns it into a package as a first step to be officially uploaded later to openupm and as long as there are no breaking changes? I might be willing to do that but what I am not going to do is to work hours on it only to have it rejected out of hand. I am concerned because I see PRs like #343 that have gone unanswered for years being as simple as editing a txt.

**Comment by Demigiant** [comment]
Woah, lot to unpack (pun not intended 😬). I'm out today, but let me get back to you all tomorrow, so I can write with more attention.

**Comment by bdovaz** [comment]
@Demigiant don't forget us!

I have been analyzing the repository and I see that you have several c# projects that have hard dependencies on specific Unity editor version assemblies.

Tasks that I see as a minimum that need to be done:
- [ ] Look if there is legacy code that no longer makes sense to continue to exist. I see folders from Unity 4, Unity 5, ... But maybe I'm wrong and it doesn't refer to anything legacy.
- [ ] Move all the source code into the folder structure of a Unity package: https://docs.unity3d.com/Manual/cus-layout.html
- [ ] Create *.asmdef files and allow Unity to compile everything, i.e. no precompiled dlls.
- [ ] Create version defines for modules such as: audio, physics, sprite, ...
- [ ] Check that windows, settings, etc still work and there are no regressions.

I see two defines: COMPATIBLE and RIGIDBODY. What is the first one for?

And if you are worried about publishing in the asset store, there is this solution for packages: https://github.com/needle-tools/hybrid-packages

**Comment by Demigiant** [comment]
> While all this is true, we have to admit upm is in a shitty state.
Using dotween as dep in another package does nothing in vanilla upm. You’d have to use openupm-cli or @mob-sakai‘s helper packages (do they even still work in 2021+?).

@krisrok You are voicing a lot of my concerns indeed. Another problem that I have with package manager (which I probably mentioned time ago in this thread) is that it always felt very alpha to me, which is one of the reasons I just give up easily when it drives my crazy

> Due to the license restriction, the com.demigiant.dotween package has been removed from OpenUPM (both from the website and the registry).

To be clear I never complained about the OpenUPM package breaking my license. I admit that initially I wasn't superhappy when I saw that there was one and I hadn't been asked about it, but I understood it and I was ok with it.
@favoyang I would also (probably) be willing to change the license a little bit in order to allow the com.demigiant.dotween OpenUPM, if you're open to discuss it, but from what I understand it's something that is too complicated to maintain (as long as I keep DOTween's main repo like this), right?

> I'm trying to avoid reading the DOTween source so that, if I decide to make my own tweening library, I won't accidentally copy the implementation.

I appreciate it @rhys-vdw, and best of luck if you go that way. When I made my first tween engine and then my second one (this one), I took the same approach with the tween engines I was using before (on Flash) so I could both avoid copying them by mistake, and create something that used its own philosophy. (Then one Unity tween engine that I won't mention, years ago, copied code from mine and I was a little annoyed)

@bdovaz There is legacy code but it's either completely disabled or in separate classes that are not compiled with anything. The COMPATIBLE define is on unused legacy code (which fixed a problem with Unity's structs on Win apps) and RIGIDBODY is part of legacy classes, from the time when Unity "extras" (like physics, UI, etc.) were enabled via defines instead than by auto-rewriting some loose scripts when importing DOTween (a big change I made a while ago because I really didn't like that I was forcing extra defines on users' projects).

The other changes are pretty big though. Making DOTween into loose scripts would break a lot of things, and I'm still not convinced: I'd really like to continue keeping the DLL approach I have now (yes, sorry, I'm still not convinced by the importance of UPM, also because I freelance to survive and I never worked on a project that was using it 👀—which doesn't mean I don't believe it's cool, quite the contrary, but it still puzzles me, probably because I associate it to the mess that is Unity's Package Manager)

**Comment by Demigiant** [comment]
P.S. To you all: I'm sorry if I continue to be so reluctant, and I hope I don't come out as dismissive. I appreciate all that you are saying (and those checkboxes looked great, @bdovaz, I love checkboxes)

**Comment by bdovaz** [comment]
@Demigiant If I create a PR that creates the package structure with dlls and coexisting with your current development, would you be willing to merge it? The idea would be to merge more or less what yoshida did but making everything work as it should.

**Comment by favoyang** [comment]
@Demigiant, thanks for the reply.

Demigiant has never pushed us to remove the published fork on OpenUPM. It's just safe for OpenUPM to remove the unlicensed package first, then discuss how to make it work like changing the license.

An OSI-approved license from [spdx.org/licenses/](https://spdx.org/licenses/) is encouraged (a dual license is okay).

However, you can keep the custom license as long as it conforms to
- Allows redistribution as open-source software.
- Allows modifications for adding necessary asmdef files, package.json, docs, and related unity meta files.

It would be good if you approve adding UPM support to this repository directly. Then OpenUPM will track and publish from this repository, which drives you happier I guess. Otherwise, a community-maintained up-to-date fork with UPM support will be used, like before.

**Comment by favoyang** [comment]
> While all this is true, we have to admit upm is in a shitty state.
> Using dotween as dep in another package does nothing in vanilla upm. You’d have to use openupm-cli or @mob-sakai‘s helper packages (do they even still work in 2021+?).

> @krisrok You are voicing a lot of my concerns indeed. Another problem that I have with package manager (which I probably mentioned time ago in this thread) is that it always felt very alpha to me, which is one of the reasons I just give up easily when it drives my crazy

It's Unity's decision to make the PackageManager (PacMan) strictly limit to the scope list, and DO NOT search packages from installed scoped registries. The scoped registry is designed to work with only a few scopes, pretty much just `com.yourcompany`. 

Well, a public registry like OpenUPM is an accident for Unity. OpenUPM manages many scopes instead of just one. The limitation of PacMan is not friendly for us. It troubles every user. Eventually, I have to write the OpenUPM-CLI for crossing it.

Because of the Unity Package Guideline, it's not allowed to write a better packman GUI with a search scope behavior. So you have to use the CLI or follow the instruction from OpenUPM to update the scope list yourself (OpenUPM provides a list of scopes needed for a given package).

The barrier has been there for years. They know it, we know it.

**Comment by bdovaz** [comment]
@Demigiant @favoyang PR created: https://github.com/Demigiant/dotween/pull/586

I think I'm going to make a lot of people happy (myself included), or at least that's my intention.

**Comment by Demigiant** [comment]
Wait wait wait, @bdovaz, we were still in the (very slow, I admit) discussion phase! Like @favoyang suggested, other than modifying this repo, I would prefer to modify the license in order to allow UPM modifications and have it stored somewhere else. One of the reasons is that I really want to keep this main DOTween version compatible with older Unity versions (both because I know people that still use Unity 5, because I have a fixation on making stuff as retro-compatible as possible, and because there is basically no performance advantage in upgrading it to later Unity versions).

@bdovaz Since you already started working on it (it scared me that it was so sudden, but I also appreciate it), maybe you could manage the OpenUPM fork? That is, if we all agree and it sounds good. What do you think, @favoyang? Let me know: since we would make it official I'll also modify DOTween's credits to indicate who's managing that side.

If that all sounds good to you, I'll post a modified license here so you can let me know if it's ok (but after the weekend: I'm crazy busy because I'm closing the UI for a game and also having to deal with some hospital things—I'm ok, it's not that—plus I absolutely need a couple hours or more to fix a fringe DOTween bug that I thought I had fixed and that recently re-appeared—the indexOutOfRange exception, which is caught and dealt with but I don't want it to happen at all, obviously).

Let me know, and bear with my slowness a little more

**Comment by Demigiant** [comment]
P.S. just so you know: I always use the latest non-beta version of Unity when possible. I'm not one of the people using Unity 5, but still :D

**Comment by favoyang** [comment]
> @bdovaz Since you already started working on it (it scared me that it was so sudden, but I also appreciate it), maybe you could manage the OpenUPM fork? That is, if we all agree and it sounds good. What do you think, @favoyang? Let me know: since we would make it official I'll also modify DOTween's credits to indicate who's managing that side.

It all sounds okay to me. This is just more work for @bdovaz to keep track of the upstream changes since he modified the directory structure. Git is not good at tracking files that moved, I assume it can not be automatically as a `git pull upstream`.

I'm also curious about the new license. But I'm positive that we can find something that works.

**Comment by bdovaz** [comment]
@Demigiant If the problem is 5.x compatibility I can multitarget the DOTween and DOTweenEditor projects to .NET Framework as before and to .NET Standard 2.0 for the package.

I "think" that is the only thing to modify to get it working again in versions prior to 2019.x (including 5.x).

**My idea was to do this PR work so that it is in the official repository, not having to maintain a fork myself**, the community wants to be maintained in an official way. Actually if we set it up well with CI it would be automatic and we wouldn't have to maintain much more since from what I see the development of DOTween (at least the free part) is quite stopped, I don't see much movement of commits.

This PR also greatly improves the flow of contributions because now we have with a workflow all the validation of the compilation of the package in multiple versions of Unity in an automated way.

I have no problem to help you to make the package flow as automatic as possible so that it has almost no maintenance but as long as it is done in the official repository.

**Comment by rhys-vdw** [comment]
What is the rationale for have a separate fork? I don't believe there are any downsides to supporting UPM in the main package. So long as edit code is under folders named `Editor`, Unity will just ignore the asmdefs and bundle everything into a single assembly (as was the way at the time).

**Comment by bdovaz** [comment]
> What is the rationale for have a separate fork? I don't believe there are any downsides to supporting UPM in the main package. So long as edit code is under folders named `Editor`, Unity will just ignore the asmdefs and bundle everything into a single assembly (as was the way at the time).

Actually the repository is not assembled this way because everything is precompiled except the modules that are accompanied by an asmdef file. That is to say, in the editor folder there is no source code, only assemblies.

But regardless of what you say, if they can coexist the current version "asset mode" and the new version "package mode" there should be no problem and we can integrate everything in this repository and that's what I've been thinking since I started working on the PR.

**Comment by bdovaz** [comment]
@Demigiant I have already corrected the fact that it generates the assemblies with the correct target framework to maintain backward compatibility, I have added more commits to the PR and I have edited the PR post adding what I have modified.

Also you can test or give me access if you want to give you a hand to check the DOTween Pro part in case there is something that could have been affected by these changes.

I have reviewed the following projects:

DOTween
DOTweenEditor
DOTweenUpgradeManager
DOTween_LooseScripts

**Comment by Demigiant** [comment]
Hey! I'm spending waaaaay more time in hospitals than I had estimated, so I'm a little overwhelmed right now. Since this is such a major change, I'll need at least two free days to check your merge request, @bdovaz, which I thought I would find max next week, but I might even need to go to the one after. I'm very conservative with my repos (until now they were more like open source that you can download but that I maintain by myself, which is my way to manage my mental health both because I'm very obsessive and because of how delicate it all is), so let's wait until I do that, but in the meantime I'm convincing myself to make this big change happen, and thank you for all the work :)

**Comment by VasylRomanets-MoonActive** [comment]
Get better soon, @Demigiant!

**Comment by Demigiant** [comment]
@VasylRomanets-MoonActive Thank you! I'm good, I'm just taking care of my aunt that is sick, but every day something new happens and the hospital(s) are far

**Comment by Demigiant** [comment]
Super late as usual with this issue, but things are finally better and I'm dealing with a queue of support before moving to this. I tried to create two separate repos that I would maintain independently but realized it was a very bad idea, so I'm moving min version of DOTween to Unity 2018 and trying to fully merge @bdovaz work.

**Comment by mitay-walle** [comment]
great news. also wanna mention:
to install current @bdovaz package version from git, you should use this text:

https://github.com/bdovaz/dotween.git?path=/Package#feature/dotween-package

**Comment by Laumania** [comment]
Hope our aunt are ok and getting better.

I really look forward to a upm/package.json version of DOTween, as Unity now seem to support git dependencies (as I read it, haven't tried it yet https://docs.unity3d.com/Manual/upm-git.html) as it will makes it a lot easier for modders of my game to get started. 

:)

**Comment by Laumania** [comment]
Let me correct myself, package.json DO NOT support git dependencies...jesus Unity.

> Note: You cannot specify Git dependencies
 in a [package.json](https://docs.unity3d.com/2021.3/Documentation/Manual/upm-manifestPkg.html) file because the Package Manager does not support Git dependencies between packages. It only supports Git dependencies for projects, so you can only declare Git dependencies in the project’s [manifest.json](https://docs.unity3d.com/2021.3/Documentation/Manual/upm-manifestPrj.html) file.
 
 https://docs.unity3d.com/2021.3/Documentation/Manual/upm-git.html
 
 Either way, a UPM version of DOTween would be nice still :)

**Comment by rhys-vdw** [comment]
@Laumania it's off topic, but if you want nested dependencies they can be installed via OpenUPM. So you can still have your mod framework on OpenUPM and have your users install it, including its deps, using the openupm CLI tool.

Or simply provide them a Unity project scaffold that includes git dependencies in its manifest.json.

**Comment by Laumania** [comment]
> @Laumania it's off topic, but if you want nested dependencies they can be installed via OpenUPM. So you can still have your mod framework on OpenUPM and have your users install it, including its deps, using the openupm CLI tool.
> 
> Or simply provide them a Unity project scaffold that includes git dependencies in its manifest.json.

@rhys-vdw 

The problem is, it makes things a lot more complicated for users who wants to mod my game and I want that to be easy.
I actually ended up doing it differently, as I figured out after looking at the license, that I am actually allowed to distribute the standard/free version of DOTween, as I'm not making any changes to it. http://dotween.demigiant.com/license.php

So I have included it in my Mod Tools instead, which makes it MUCH easier for users to build mods.

I only need the DOTween for the mod to build, it's not actually included in the mod, when mod is loaded up in the game its actually using the DOTween within the game.

Sorry for off-topic :P

**Comment by pawelsalwa** [comment]
Hello,
Is there any update on this? Is the feature rejected or still a possibility?

**Comment by callum-rose-red-engine** [comment]
Bump on any update, would be super useful :)

**Comment by Demigiant** [comment]
Ahoy,

I'm sorry but I never know how to answer this. In theory I would go for "maybeee", because every time I try to get to it I get derailed by doing other DOTween features (because everything needs to be perfect before this process), and then it gets postponed once again because I'm overloaded.

As a secondary but important note, like @Laumania said you can totally redistribute DOTween with your projects/etc. (as long as you redistribute it "as-is"), so that should help in some cases in the meantime.

**Comment by celikomerdev** [comment]
> Ok I created an external DOTween package and tried until now to make it work, but I finally had to ragequit :B I managed to get all references correctly, but the problem is that I also need to modify files via the editor, and it's a complete mess. I experimented but the lack of documentation is too great, so I'll postpone this for the day Unity decides to make Package Manager a little more open, sigh :|
> 
> In the meantime, [here's v1.2.225](http://dotween.demigiant.com/tmp/DOTween_1_2_225.zip), where I added a button in DOTween Utility Panel to create the asmdef files. It creates one for the modules and one for DOTweenPro loose scripts (if present). I managed to create DOTweenPro's directly in the DOTweenPro folder by manually adding references to the asmdef json when it's created.

Hi there,
As i see you are modifying source files with "#if true // MODULE_MARKER" flag, but this workflow seems wrong,
You can add global define symbols to the project settings such as "DOTWEEN_AUDIO" and change your source code flag with "#if DOTWEEN_AUDIO" and you would never have to change source code again.
You can use utility panel to add or remove global define symbols to the project settings.
PlayerSettings.SetScriptingDefineSymbolsForGroup();

**Comment by Demigiant** [comment]
> As i see you are modifying source files with "#if true // MODULE_MARKER" flag, but this workflow seems wrong,
> You can add global define symbols to the project settings such as "DOTWEEN_AUDIO" and change your source code flag with "#if DOTWEEN_AUDIO" and you would never have to change source code again.

That's how older DOTween versions worked, but I really didn't like it so I made a major refactoring to auto-rewrite DOTween's files instead of a user's project defines. My philosophy is that a plugin/asset should strive to change a project as little as possible, and adding lots of defines (especially considering how cramped Unity's define box is, and that someone might have their own defines in there) felt way too intrusive to me 😬

**Comment by ErnSur** [comment]
The UPM package needs to be immutable.

There is a way to create a UPM package with "module" dependent code compilation without using project defines.
The [Version defines list](https://docs.unity3d.com/Manual/ScriptCompilationAssemblyDefinitionFiles.html#define-symbols) in the assembly definition asset can be used for that. This feature allows you to enable defines in an assembly if a specific package is present in a project. 

- Each DOTween module could be made into a separate UPM package.
- To enable a module you would simply need to add the module package to the project. It doesn't matter if the module was added using git, local path, or registry.
- After adding a module package the assembly definitions from the main package will recompile with additional defines because of version defines list

---

I can imagine how splitting this library into multiple packages can be troublesome during the delivery and installation process.
To streamline it the following approach could be used:

In the Unity Editor window for module management user could choose to 
- Add a module package as a local package inside the Packages folder
    - Simply copy and paste a module package from the main package to a project Packages folder.
    - Module packages can be stored in the main package hidden directory (just like package samples, this will prevent modules from compilation)
    - Module packages stored in the main package could be in a tarball format
- Add a module package as a git dependency (install a package using a GitHub repo https address)

[API](https://docs.unity3d.com/ScriptReference/PackageManager.Client.html) to add and remove project packages.

**Comment by laicasaane** [comment]
> My philosophy is that a plugin/asset should strive to change a project as little as possible, and adding lots of defines (especially considering how cramped Unity's define box is, and that someone might have their own defines in there) felt way too intrusive to me 

@Demigiant From the perspective of an advanced developer, this is no problem at all.

> Enabling a module will simply import the relevant package into the project.

The purpose of this issue is to avoid polluting the main Assets folder. I can live with a bunch of compilation symbols.

**Comment by ErnSur** [comment]
>The purpose of this issue is to avoid polluting the main Assets folder. I can live with a bunch of compilation symbols.

You misunderstood, importing a package has nothing to do with the Assets folder.
In the solution that I proposed a module is a separate UPM package, so you can import it in any way you want: local, git or registry.

**Comment by ErnSur** [comment]
I updated my previous comment with a better explanation of the idea.

**Comment by rhys-vdw** [comment]
@Demigiant 

> My philosophy is that a plugin/asset should strive to change a project as little as possible, and adding lots of defines (especially considering how cramped Unity's define box is, and that someone might have their own defines in there) felt way too intrusive to me 😬

See how this works in section "Defining symbols based on project packages":
https://docs.unity3d.com/2020.1/Documentation/Manual/ScriptCompilationAssemblyDefinitionFiles.html

Basically the DOTween module would just set a flag for its own compilation based on whether another module was included. This is how UniTask flags its DOTween support on. In DoTween's case it could set a flag for e.g. UNITY_UI etc.

**Comment by LarsBlaabjerg** [comment]
Sorry for barging into the discussion. Just chiming in with a few extra details.

I don't think it it would be necessary to split the library into separate packages.

I think it was mentioned previously, but by using assembly definitions and the "Define Constraints" feature along with the "Version Defines" to limit when the assembly is used, you would be able to create assemblies for every occasion or have conditional code in them.

The "Version Defines" feature is used to define a preprocessor symbol based on an expression using the present version of a package or unity module, or unity itself. That symbol can then be used in the code with #if, or in "Define Constraints".

![version-defines](https://user-images.githubusercontent.com/4222097/231736830-b02da8bb-2f0e-4137-bf60-b8ac21ba8071.png)
![define-constraints](https://user-images.githubusercontent.com/4222097/231736873-3136e983-9a59-4bdd-9116-6441f2dc5c63.png)

By doing this, it should be possible to remove the need to modify the source files, allowing the library to exist in an immutable package.

It should also not be necessary to add any global defines.

These assembly definition features are supported from Unity version 2019 (according to the documentation).

BTW: I think if code is in a separate assembly, it will be stripped from builds if there are no references to it from an included scene, assets in the resources folder, or from other code that is not itself stripped. So I am thinking that maybe you don't have to worry so much about your library forcing Unity to include components that it otherwise wouldn't have. I am kind of assuming this though.

**Comment by ErnSur** [comment]
@LarsBlaabjerg The need for separate packages is coming from the feature of enabling and disabling certain DOTween modules. Even with the help of version define list it would be impossible to toggle a module inside an immutable package. (In a case where all modules and a core lib are inside the same UPM package)

**Comment by LarsBlaabjerg** [comment]
> @LarsBlaabjerg The need for separate packages is coming from the feature of enabling and disabling certain DOTween modules. Even with the help of version define list it would be impossible to toggle a module inside an immutable package. (In a case where all modules and a core lib are inside the same UPM package)

But you can have several assembly definitions in the same package. True. If you need to be able to toggle available features at a whim, you would need to add global defines (which is a fine way to do it btw), but I don't understand the need to to. Why not juts let all the modules be available if the prerequisites are?

**Comment by ErnSur** [comment]
You're right, I agree. 👍

**Comment by rhys-vdw** [comment]
Yes, what @LarsBlaabjerg suggests is the standard way of doing things. And, in fact, what I suggested in the comment before his. 😉

**Comment by LarsBlaabjerg** [comment]
Yeah, I know @rhys-vdw . I just couldn't resist elaborating it a bit :-)

**Comment by snorrsi** [comment]
Hi @Demigiant ,

I understand your concern regarding not wanted to change user project files.
Sometimes it just needs to be done in sake of clarity.

What you could do make it more clean is to create a auto-generated settings file Assets/dotween.rsp
That file include DOTWEEN_ script defines as discussed earlier based on what the user chooses to enable.
If using the method below then all script defines in this files are included in all builds, including of packages.
The only this is that this file has to be included by refencing it (same thing could be done directly with Assets/csc.rsp file but that would break the simplicity of keeping all Dotween settings in a seperate file.

The Assets/dotween.rsp file has to be included either by referencing it:

A. Either by  it through Assets/csc.rsp reading
`@Assets/dotween.rsp`

Assets/csc.rsp it automatically included
You can do the same in any other .rsp file

or 

B. Adding it in Project Settings / Player / Script Compilation / Scripting Define Symbols
`@Assets/dotween.rsp`

Just as with the other method.

--

Pro: Simple to setup, only one file is written to and one setting is changed. Nothing is changed in the UPM package folder
Cons: Might get script compilation error if the referenced .rsp file does not exist. The user need to add the reference to the file or Dotween could do it for them.

**Comment by bdovaz** [comment]
Any news @Demigiant? It will be almost a year since my [original PR](https://github.com/Demigiant/dotween/pull/586).

Can I help you in any way? It's a bit frustrating to see all my work stopped for a year but I understand that everyone has their priorities 😅

**Comment by laicasaane** [comment]
@Demigiant I don't really understand why it's so hard on your part. This issue seems like an eternity, while I've been already developing and maintaining a bunch of UPM packages without problem. Truthfully, there are many compilation symbols in my project, defined either by me or by 3rd packages. But it's fine, because we have to declare them only once. But upgrading DOTween is a recurring task that begins getting into my nerves.

I might seem stubborn about keeping the `Assets` folder clean from 3rd assets. But it's much more than being clean. For example, for any 3rd asset that is compatible with UPM, I can just simply run this command `npm pack` to produce a `.tgz` file then install it as an embedded package. It helps reducing the number of `.csproj` generated, thus improve the performance of the IDE itself. Without the various projects from 3rd assets, the IDE processes only projects defined by me.

There is no way to stop Unity from generating `.csproj` for DOTween modules and that annoys both me and my IDE. Unloading unnecessary projects does help a bit, but we'll lose IntelliSense on their APIs.

I really hope this issue will be concluded in the near future.

**Comment by bdovaz** [comment]
_PR created in July 2022_

@Demigiant friendly ping!

This summer you have been active, I see several commits in DOTween and there are new releases in August of DOTween in the Asset Store.

I mean, I'd like to know why you don't show up here. There are a lot of people waiting for my work to merge.

I already said I help you with whatever you need.

**Comment by Demigiant** [comment]
@bdovaz first of all sorry again, and thank you for being friendly even if this got delayed for a very annoying amount of time. I always plan to get to this but then postpone it for one reason or another, and don't know how to explain it so I hide.

The reasons I work on other commits is that bugfixing is always my main priority, so if a bug comes out I try to jump on it asap regardless of the mess I'm in. Which is also one of the reasons I postpone your merge. After the times I tried to tackle this I know that, contrary to bugfixing and smaller quality additions which only require bursts of a few hours, I'll need at least two weeks, probably three, of constant involvement in order to do this, and those are not easy for me to find (and to be honest, when I did find them in the past I ended up working on other improvements, because admittedly I still don't see much value in this feature, so that's also my bad 👀).

That said, let me try again. I won't have enough time until the half/end of November and then I had planned to finish working on a small game, but I'll postpone that and give precedence to this. It's in my calendar now 👀

**Comment by insominx** [comment]
FWIW, I also need this for the same dependency reasons.  It seems I am far from alone on this.  Perhaps it should be a higher priority given the community response.

**Comment by TylerTemp** [comment]
Year 2024, after 6 years I'm still expecting for this feature (

(And thanks for this awesome project, I'm a DOTween Pro user)

**Comment by avinashmx7** [comment]
Hi @Demigiant! Firstly, many thanks for the wonderful plugin. I am anticipating this feature as well. Is there any update on this?

**Comment by bdovaz** [comment]
> @bdovaz first of all sorry again, and thank you for being friendly even if this got delayed for a very annoying amount of time. I always plan to get to this but then postpone it for one reason or another, and don't know how to explain it so I hide.
> 
> The reasons I work on other commits is that bugfixing is always my main priority, so if a bug comes out I try to jump on it asap regardless of the mess I'm in. Which is also one of the reasons I postpone your merge. After the times I tried to tackle this I know that, contrary to bugfixing and smaller quality additions which only require bursts of a few hours, I'll need at least two weeks, probably three, of constant involvement in order to do this, and those are not easy for me to find (and to be honest, when I did find them in the past I ended up working on other improvements, because admittedly I still don't see much value in this feature, so that's also my bad 👀).
> 
> That said, let me try again. I won't have enough time until the half/end of November and then I had planned to finish working on a small game, but I'll postpone that and give precedence to this. It's in my calendar now 👀

@Demigiant any news (almost a year from your response)? This PR has been stopped for 2 years with all the work it meant to me at the time....

Now the situation has also worsened because the temporary fork of https://github.com/yoshida190/dotween no longer exists and we have started to fail this manifest.json dependency in several projects: 

```
"com.demigiant.dotween": "https://github.com/yoshida190/dotween.git#v1.2.632-upm",
```

**Comment by chrisSayitlabs** [comment]
Is there any update on this? I'm making packages of some of our systems at the moment and we have used DOTween for a long time in some of these systems. It would be great to not have to swap to something else.

**Comment by bdovaz** [comment]
The PR has been created for 2 and a half years now without merge and with no signs of changing...

What bothers me the most @Demigiant is the hours of work dedicated to it for nothing.... With how difficult it is many times to find contributors in OSS...

For some time now I have been using a different tween library and I have removed all traces of DOTween in my projects...

**Comment by Peaj** [comment]
I have been using DoTween in the past as I really like its API. I am currently evaluating tweening solutions again and one of the most important features for me is package manager support as its just a lot easier to manage dependencies and track updates.
Its really nice to see that people are actually working on it and I hope this eventually gets resolved.

**Comment by callum-rose-red-engine** [comment]
Yeah I'm just gonna add my 2 cents here and say this is an important feature for us too. Would be amazing if we could have it!

**Comment by sprimate** [comment]
Bump to take a look at that PR.

Dependency management is important, the company I work for has a hard requirement for keeping packages in UPM.  

I recognize that you don't think of this as important as bugs, but I would argue that this is MORE important than bugs, because when we realized there was no UPM version of this package, we initially assumed that it was no longer being supported at all, and started looking for alternatives, because we didn't want to add "legacy" code into our Unity 6 project.

When actually, it's pretty regularly supported and very well-maintained, it's just missing what I would consider a core feature for a unity package in a Unity 6+ world (really since like Unity 2020, but especially 6+)

Unless this package is in a "Maintain-Only" state by the creator, where you're not adding new features cause you're living your life, which I TOTALLY understand. 

But if the goal IS for this to be a long-term solution, I'd argue that checking out that UPM PR (or at the very least, accepting it into a "beta"/"use-at-your-own-risk" branch) should be a pretty high-priority thing.

**Comment by Quawetim** [comment]
Agree with above. It's now pain in the ass from the past. We use in-house game engine package for fast prototyping and development. And most plugins and packages are at that package. Without asmdef support from DOTween we can't do whatever we want to do.

I'm also dont understand why PR dies almost 3 years.

@bdovaz, as i understand, project owner don't care anymore. Can you advise alternative? What was chosen by you?

**Comment by bdovaz** [comment]
> @bdovaz, as i understand, project owner don't care anymore.

That seems and as you can understand with the time and effort spent by me in creating the PR and see how years later is abandoned, It drives people away from contributing to open source projects... There are projects where they would like to have contributions like this and in this case @Demigiant completely ignores it...

> Can you advise alternative? What was chosed by you?

https://github.com/annulusgames/LitMotion

**Comment by Quawetim** [comment]
> It drives people away from contributing to open source projects..

Strongly agreed.

> https://github.com/annulusgames/LitMotion

Thank you. I was looking at this too.

**Comment by LittleBigMonkey** [comment]
I'm following this thread since the beginning... And nothing has changed.
It might be the right moment to move on to a new tool.


> https://github.com/annulusgames/LitMotion

Thank you, @bdovaz. That looks very promising, and the claimed performance are even better than DotTween’s.

**Comment by insominx** [comment]
@Demigiant Please update us.  If this will never happen, let us know so that we can move on.

**Comment by TylerTemp** [comment]
> https://github.com/annulusgames/LitMotion

Thank you, @bdovaz  . Im moving to `LitMotion` too even I have DOTween Pro purchased.

**Comment by Ayfel** [comment]
Just a friendly bump in here as this is an important feature we would appreciate
